Add unsealing capability to `mc-sgx-tservice`

nick-mobilecoin opened the pull request.

nick-mobilecoin committed:
Add unsealing capability to `mc-sgx-tservice`

nick-mobilecoin commented:
Current dependencies on/for this PR:
* main
  * **PR #125** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #126** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/126" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #127** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/127" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/127?utm_source=stack-comment).

nick-mobilecoin commented on deny.toml:
```diff
@@ -36,8 +36,16 @@ deny = [
 ]
-skip = []
-skip-tree = []
+skip = [
```
Not fond of this, a little surprised there isn't a way to allow path dependencies and only check their transitive dependencies.
This may all go away though if #125 collapses to opaque types as the common test logic would no longer be needed in `mc-sgx-tservice-types`

nick-mobilecoin replied:
gone now

nick-mobilecoin commented:
Current dependencies on/for this PR:
* main
  * **PR #125** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #126** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/126" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #127** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/127" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/127?utm_source=stack-comment).

nick-mobilecoin commented:
Current dependencies on/for this PR:
* main
  * **PR #125** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #126** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/126" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #127** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/127" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/127?utm_source=stack-comment).

jcape commented on tservice/src/seal.rs:
```diff
@@ -71,6 +71,77 @@ impl<T: AsRef<[u8]>> Seal<Vec<u8>> for UnsealedData<T> {
+        }
+    }
+
+    fn unseal(&self) -> Result<UnsealedData<Vec<u8>>> {
```
This necessarily means we're going to need an `Unseal::unseal<T>(&self, &dest: UnsealedData<T>)`, but...

```suggestion
    #[cfg(feature = "alloc")]
    fn unseal_vec(&self) -> Result<UnsealedData<Vec<u8>>> {
```

nick-mobilecoin replied:
I thought alloc  as a feature was only needed for things outside of the enclave.

If it's true we don't need separate alloc logic:
- should `Unsealed` be converted to a `Vec<u8>` instead of a generic?
- should the `Unseal` go away and be methods on the `Unsealed` struct?

jcape commented on tservice/src/seal.rs:
```diff
@@ -71,6 +71,77 @@ impl<T: AsRef<[u8]>> Seal<Vec<u8>> for UnsealedData<T> {
+            _ => mac.as_mut_ptr(),
+        };
+
+        let mut data_length = data_length as u32;
+        let mut mac_length = mac_length as u32;
```
That these are being passed as `mut` to the FFI binding is giving me pause. I feel like we should return an internal error if `mac_length` or `data_length` has changed...

nick-mobilecoin replied:
I think this goes back to the common C api, of here is my buffer and it has room for X bytes, then the function fills the buffer with Y bytes replacing the size with the _used_ size.
```
     *      p_additional_MACtext_length - [IN/OUT] pointer to length of the plaintext data stream in bytes
     *                             If there is not additional data, this parameter should be zero.
```
This may be a failure on my part to [resize](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.resize) with the values afterward.

Or are you thinking that they shouldn't change since previously we got the needed length?

nick-mobilecoin replied:
https://github.com/mobilecoinfoundation/sgx/pull/127/commits/18bf2d20d0b8caf3d0dd6c81560ce40336c2faca added logic to error if the sizes differ from what SGX told us originally.

jcape commented on tservice/test_utils/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+readme = "README.md"
+repository = "https://github.com/mobilecoinfoundation/sgx"
+description = "Test utilities for the rust wrappers of the `sgx_tservice` library."
+
+categories = ["development-tools::testing"]
+keywords = ["sgx", "test"]
```
I'm ambivalent about this... If we're going to go down the "grouping" route I'd like to 1. make more groups and 2. make everything look the same.

nick-mobilecoin replied:
resolving for now as this is in other cargo.toml files and should have a dedicated normalization effort

jcape reviewed (changes requested):
I'd like to chat about the tservice test_utils crate, and whether we should just make it a module in tservice that's feature-disabled by default?

Also we should gate `Unseal::unseal()` (as it currently exists) on "alloc", and either do one that writes to a `dest: UnsealedData<T>`, or maybe is just `From<SealedData<T>> for UnsealedData<T>` :-)

nick-mobilecoin committed:
Add error handling if unsealed data sizes deviate

Previously in `Unseal::unseal()` the data size and mac text sizes were
not inspected after the call to `sgx_unseal_data`.  Now these values are
checked and if they are different than the sizes originally provided by
SGX an error is returned.

nick-mobilecoin committed:
Remove the `tservice/test_utils` crate

The local only `tservice/test_utils` crate has been converted to the
module `mc-sgx-tservice-types::test_utils`.  This module is hidden
behind the `test-utils` feature in `mc-sgx-tservice-types`.

jcape commented on tservice/src/seal.rs:
```diff
@@ -90,10 +90,61 @@ impl<T: AsRef<[u8]> + core::default::Default> SealedBuilder<T> {
     }
 }
 
+pub trait Unseal<T> {
+    /// The length (in bytes) needed to hold the decrypted text
+    fn decrypted_text_len(&self) -> Result<usize>;
+
+    /// Unseal the data in `self`
+    fn unseal(&self) -> Result<T>;
+}
```
I'm not sure I like putting the type param on this trait---it feels like it's really only capable of being used for `Vec` and constant arrays.

My preference would be to have something like this:

```rust
pub trait Unseal {
    fn decrypted_text_len(&self) -> Result<usize>;
    fn unseal(&self, dest: &mut [u8]) -> Result<usize>;

    fn unseal_vec(&self) -> Result<Vec<u8>> {
        let mut retval = vec![0u8; self.decrypted_text_len()];
        let retval_len = self.unseal(&mut retval)?;
        retval.truncate(retval_len);
        Ok(retval)
    }
}
```

The common case (give me the data as a `Vec`) is handled, but not at the expense of foreclosing the caller into using the heap or trying to guess at the size of the data from the size of the data in `Sealed<>`

nick-mobilecoin replied:
Reworked to have an `unseal` and `unseal_to_vec` method.
The trait is generic over `AsRef<[u8]>` which should match the scope of `Sealed`

jcape reviewed (commented):
I'm ambivalent about Unseal<T>, do you want to chat about it?

nick-mobilecoin committed:
Change `mc-sgx-tservice::Unseal::unseal` to return &mut [u8]

Previously `mc-sgx-tservice::Unseal::unseal` returned a Vec.  Now
`mc-sgx-tservice::Unseal::unseal` returns a `&mut [u8]`.  A new method,
`mc-sgx-tservice::Unseal::unseal_to_vec` has been added that outputs a
Vec.

awygle commented on tservice/src/seal.rs:
```diff
@@ -90,10 +90,74 @@ impl<T: AsRef<[u8]> + core::default::Default> SealedBuilder<T> {
+        // SGX interface at the top of the function for the required sizes.
+        // If the sizes are different than we have unexpected behavior.
+        if data_length != data_length_u32 as usize || mac_length_u32 != 0 {
+            return Err(Error::Unexpected);
```
It may be worth exercising this codepath in the tests.

nick-mobilecoin replied:
Unfortunately this logic we can't exercise without coming up with some kind of framework to run in the enclave.  
If one tries to compile a test for this crate it will fail to link up because it will try to pull in std for the test run, but the enclave code is no_std and when it links in the sgx libs it will be missing the allocator, panic handler, etc.

awygle replied:
Ah, that's irritating. In that case can you file an issue for future work about figuring out how to execute tests of this type? We're going to run into more of them in the future, I'm sure.

nick-mobilecoin replied:
Created, #135

awygle commented on tservice/src/seal.rs:
```diff
@@ -90,10 +90,74 @@ impl<T: AsRef<[u8]> + core::default::Default> SealedBuilder<T> {
+    fn unseal<'a>(&self, buffer: &'a mut [u8]) -> Result<&'a mut [u8]> {
+        let data_length = self.decrypted_text_len()?;
+        if buffer.len() < data_length {
+            return Err(Error::InvalidParameter);
```
It may be worth exercising this codepath in the tests.

nick-mobilecoin replied:
see above

awygle reviewed (changes requested):
LGTM, just needs to cover failure cases in Unseal testing.
